File: pyscore/bmc/link.py
#  Copyright 2020 ICON Foundation
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import math

from iconservice import Address, IconScoreDatabase

from .relay import BTPRelays, BTPRelay
from ..lib import BTPAddress, BTPException
from ..lib.icon import PropertiesDB, IterableDictDB, remove_from_array_db


class BTPLink(PropertiesDB):
    def __init__(self, prefix, db: IconScoreDatabase, addr: BTPAddress) -> None:
        self.addr = addr  # primary key
        self.rx_seq = 0
        self.tx_seq = 0

        self.block_interval_src = 0
        self.block_interval_dst = 0
        self.max_aggregation = 10  # TODO define default max_aggregation
        self.delay_limit = 3  # if delay_limit < 3, too sensitive

        self.relay_idx = 0
        self.rotate_height = 0
        self.rx_height = 0  # initialize with BMC.block_height
        self.rx_height_src = 0  # initialize with BMV._offset

        # next-relays
        self.next_relays_height = 0
        self.next_relays_term = 0  # 0: disable next-relays, 43200: number of block daily

        # SACK
        self.sack_term = 0  # 0: disable sack
        self.sack_next = 0
        self.sack_height = 0
        self.sack_seq = 0

        if prefix is None:
            prefix = str(addr)
        super().__init__(prefix, db)
        self.relays = BTPRelays(self._sub_prefix("relays"), self._base_db)
        self.reachable = super()._array_db("reachable", str)

        self.next_relays_add = super()._array_db("next_relays_add", Address)
        self.next_relays_remove = super()._array_db("next_relays_remove", Address)

    def _scale(self) -> float:
        if self.block_interval_src < 1 or self.block_interval_dst < 1:
            return 0
        return self.block_interval_src / self.block_interval_dst

    def rotate_term(self) -> int:
        scale = self._scale()
        if scale > 0:
            return math.ceil(self.max_aggregation / self._scale())
        else:
            return 0

    def _rotate(self, rotate_count: int, base_height: int) -> BTPRelay:
        rotate_term = self.rotate_term()
        if rotate_term > 0 and rotate_count > 0:
            self.rotate_height = base_height + rotate_term
            self.relay_idx += rotate_count
            if self.relay_idx >= len(self.relays):
                self.relay_idx = self.relay_idx % len(self.relays)
        return self.relays.at(self.relay_idx)

    def rotate(self, current_height: int, relay_msg_height: int, has_msg: bool = False) -> BTPRelay:
        rotate_term = self.rotate_term()
        if rotate_term > 0:
            if has_msg:
                guess_height = self.rx_height + math.ceil((relay_msg_height - self.rx_height_src) / self._scale()) - 1
                if guess_height > current_height:
                    guess_height = current_height
                rotate_count = math.ceil((guess_height - self.rotate_height)/rotate_term)
                if rotate_count < 0:
                    rotate_count = 0
                base_height = self.rotate_height + ((rotate_count-1) * rotate_term)
                skip_count = math.ceil((current_height - guess_height)/self.delay_limit) - 1
                if skip_count > 0:
                    rotate_count += skip_count
                    base_height = current_height
                self.rx_height = current_height
                self.rx_height_src = relay_msg_height
                return self._rotate(rotate_count, base_height)
            else:
                rotate_count = math.ceil((current_height - self.rotate_height)/rotate_term)
                base_height = self.rotate_height + ((rotate_count-1) * rotate_term)
                return self._rotate(rotate_count, base_height)
        else:
            return None

    def add_relay(self, addr: Address):
        if self.next_relays_term > 0:
            if addr in self.relays or addr in self.next_relays_add:
                raise BTPException("already registered relay")
            if remove_from_array_db(self.next_relays_remove, addr) < 0:
                self.next_relays_add.put(addr)
        else:
            if addr in self.relays:
                raise BTPException("already registered relay")
            self.relays[addr] = self.relays.new_relay(addr)

    def remove_relay(self, addr: Address, current_height: int):
        if self.next_relays_term > 0:
            if addr not in self.relays:
                raise BTPException("not registered relay")
            if addr in self.next_relays_remove:
                raise BTPException("already unregistered relay")
            if remove_from_array_db(self.next_relays_add, addr) < 0:
                self.next_relays_remove.put(addr)
        else:
            if addr not in self.relays:
                raise BTPException("not registered relay")
            self.relays.remove(addr)
            if 0 < len(self.relays) <= self.relay_idx:
                self._rotate(1, current_height)

    def get_next_relays(self) -> list:
        relays = []
        remove_relays = []
        for relay in self.next_relays_remove:
            remove_relays.append(relay)
        for relay in self.relays.keys():
            if relay not in remove_relays:
                relays.append(relay)
        for relay in self.next_relays_add:
            relays.append(relay)
        return relays

    def next_relays(self, current_height: int):
        # TODO how to trigger next_relays?
        if current_height >= self.next_relays_height:
            for i in range(len(self.next_relays_remove)):
                self.relays.remove(self.next_relays_remove.pop())
            for i in range(len(self.next_relays_add)):
                relay = self.next_relays_add.pop()
                self.relays[relay] = self.relays.new_relay(relay)
            self.next_relays_height = self.next_relays_height + self.next_relays_term


class BTPLinks(IterableDictDB):

    def __init__(self, key, db: IconScoreDatabase):
        super().__init__(key, db, BTPLink)

    def raw_to_object(self, raw):
        return self.new_btp_link(BTPAddress.from_string(raw))

    def object_to_raw(self, obj):
        return obj

    def __getitem__(self, key: object) -> BTPLink:
        return super().__getitem__(key)

    def remove(self, key) -> None:
        v = self.__getitem__(key)
        if v is not None:
            v.remove_all()
            super().remove(key)

    def new_btp_link(self, addr: BTPAddress) -> BTPLink:
        return BTPLink(self._sub_prefix(str(addr)), self.base_db, addr)

File: javascore/bmc/src/main/java/com/iconloop/btp/bmc/Link.java
/*
 * Copyright 2021 ICON Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.iconloop.btp.bmc;

import com.iconloop.btp.lib.BTPAddress;
import score.ByteArrayObjectWriter;
import score.Context;
import score.ObjectReader;
import score.ObjectWriter;
import scorex.util.ArrayList;

import java.math.BigInteger;
import java.util.List;

public class Link {
    //TODO define default max_aggregation
    public static final int DEFAULT_MAX_AGGREGATION = 10;
    //TODO define default delay_limit, if delay_limit < 3, too sensitive
    public static final int DEFAULT_DELAY_LIMIT = 3;

    private Relays relays;

    //
    private BTPAddress addr; //primary key
    private BigInteger rxSeq;
    private BigInteger txSeq;
    //

    private int blockIntervalSrc;
    private int blockIntervalDst;
    private int maxAggregation = DEFAULT_MAX_AGGREGATION;
    private int delayLimit = DEFAULT_DELAY_LIMIT;

    private int relayIdx;
    private long rotateHeight;
    private long rxHeight; //initialize with BMC.block_height
    private long rxHeightSrc; //initialize with BMV._offset

    private int sackTerm; //0: disable sack
    private long sackNext;
    private long sackHeight;
    private BigInteger sackSeq;

    //with suffix("reachable") ArrayDB<String>
    private List<BTPAddress> reachable;

    private double scale() {
        if (blockIntervalSrc < 1 || blockIntervalDst < 1) {
            return 0;
        } else {
            return (double)blockIntervalSrc/(double)blockIntervalDst;
        }
    }

    public int rotateTerm() {
        double scale = scale();
        if (scale > 0) {
            return (int)StrictMath.ceil((double) maxAggregation / scale);
        } else {
            return 0;
        }
    }

    public Relay rotate(long currentHeight, long msgHeight, boolean hasMsg) {
        long rotateTerm = rotateTerm();
        if (rotateTerm > 0) {
            int rotateCnt;
            long baseHeight;
            if (hasMsg) {
//                long guessHeight = rxHeight + Util.ceilDiv((currentHeight - rxHeightSrc), scale()) - 1;
                long guessHeight = rxHeight + (long)StrictMath.ceil((double) (currentHeight - rxHeightSrc) / scale()) - 1;
                if (guessHeight > currentHeight) {
                    guessHeight = currentHeight;
                }

//                rotateCnt = (int)Util.ceilDiv(guessHeight - rotateHeight, rotateTerm);
                rotateCnt = (int)StrictMath.ceil((double)(guessHeight - rotateHeight)/(double)rotateTerm);
                if (rotateCnt < 0) {
                    rotateCnt = 0;
                }
                baseHeight = rotateHeight + ((rotateCnt - 1) * rotateTerm);
//                int skipCnt = (int)Util.ceilDiv((currentHeight - guessHeight), delayLimit) - 1;
                int skipCnt = (int)StrictMath.ceil((double)(currentHeight - guessHeight)/(double)delayLimit) - 1;
                if (skipCnt > 0) {
                    rotateCnt += skipCnt;
                    baseHeight = currentHeight;
                }
                rxHeight = currentHeight;
                rxHeightSrc = msgHeight;
            } else {
//                rotateCnt = (int)Util.ceilDiv((currentHeight - rotateHeight), rotateTerm);
                rotateCnt = (int)StrictMath.ceil((double)(currentHeight - rotateHeight)/(double)rotateTerm);
                baseHeight = rotateHeight + ((rotateCnt - 1) * rotateTerm);
            }
            if (rotateCnt > 0) {
                rotateHeight = baseHeight + rotateTerm;
                relayIdx += rotateCnt;
                if (relayIdx >= relays.size()) {
                    relayIdx = relayIdx % relays.size();
                }
            }
            return relays.getByIndex(relayIdx);
        } else {
            return null;
        }
    }

    public BTPAddress getAddr() {
        return addr;
    }

    public void setAddr(BTPAddress addr) {
        this.addr = addr;
    }

    public BigInteger getRxSeq() {
        return rxSeq;
    }

    public void setRxSeq(BigInteger rxSeq) {
        this.rxSeq = rxSeq;
    }

    public BigInteger getTxSeq() {
        return txSeq;
    }

    public void setTxSeq(BigInteger txSeq) {
        this.txSeq = txSeq;
    }

    public int getBlockIntervalSrc() {
        return blockIntervalSrc;
    }

    public void setBlockIntervalSrc(int blockIntervalSrc) {
        this.blockIntervalSrc = blockIntervalSrc;
    }

    public int getBlockIntervalDst() {
        return blockIntervalDst;
    }

    public void setBlockIntervalDst(int blockIntervalDst) {
        this.blockIntervalDst = blockIntervalDst;
    }

    public int getMaxAggregation() {
        return maxAggregation;
    }

    public void setMaxAggregation(int maxAggregation) {
        this.maxAggregation = maxAggregation;
    }

    public int getDelayLimit() {
        return delayLimit;
    }

    public void setDelayLimit(int delayLimit) {
        this.delayLimit = delayLimit;
    }

    public int getRelayIdx() {
        return relayIdx;
    }

    public void setRelayIdx(int relayIdx) {
        this.relayIdx = relayIdx;
    }

    public long getRotateHeight() {
        return rotateHeight;
    }

    public void setRotateHeight(long rotateHeight) {
        this.rotateHeight = rotateHeight;
    }

    public long getRxHeight() {
        return rxHeight;
    }

    public void setRxHeight(long rxHeight) {
        this.rxHeight = rxHeight;
    }

    public long getRxHeightSrc() {
        return rxHeightSrc;
    }

    public void setRxHeightSrc(long rxHeightSrc) {
        this.rxHeightSrc = rxHeightSrc;
    }

    public int getSackTerm() {
        return sackTerm;
    }

    public void setSackTerm(int sackTerm) {
        this.sackTerm = sackTerm;
    }

    public long getSackNext() {
        return sackNext;
    }

    public void setSackNext(long sackNext) {
        this.sackNext = sackNext;
    }

    public long getSackHeight() {
        return sackHeight;
    }

    public void setSackHeight(long sackHeight) {
        this.sackHeight = sackHeight;
    }

    public BigInteger getSackSeq() {
        return sackSeq;
    }

    public void setSackSeq(BigInteger sackSeq) {
        this.sackSeq = sackSeq;
    }

    public List<BTPAddress> getReachable() {
        return reachable;
    }

    public void setReachable(List<BTPAddress> reachable) {
        this.reachable = reachable;
    }

    public Relays getRelays() {
        return relays;
    }

    public void setRelays(Relays relays) {
        this.relays = relays;
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("Link{");
        sb.append("relays=").append(relays);
        sb.append(", addr=").append(addr);
        sb.append(", rxSeq=").append(rxSeq);
        sb.append(", txSeq=").append(txSeq);
        sb.append(", blockIntervalSrc=").append(blockIntervalSrc);
        sb.append(", blockIntervalDst=").append(blockIntervalDst);
        sb.append(", maxAggregation=").append(maxAggregation);
        sb.append(", delayLimit=").append(delayLimit);
        sb.append(", relayIdx=").append(relayIdx);
        sb.append(", rotateHeight=").append(rotateHeight);
        sb.append(", rxHeight=").append(rxHeight);
        sb.append(", rxHeightSrc=").append(rxHeightSrc);
        sb.append(", sackTerm=").append(sackTerm);
        sb.append(", sackNext=").append(sackNext);
        sb.append(", sackHeight=").append(sackHeight);
        sb.append(", sackSeq=").append(sackSeq);
        sb.append(", reachable=").append(reachable);
        sb.append('}');
        return sb.toString();
    }

    public static void writeObject(ObjectWriter writer, Link obj) {
        obj.writeObject(writer);
    }

    public static Link readObject(ObjectReader reader) {
        Link obj = new Link();
        reader.beginList();
        obj.setAddr(reader.readNullable(BTPAddress.class));
        obj.setRxSeq(reader.readNullable(BigInteger.class));
        obj.setTxSeq(reader.readNullable(BigInteger.class));
        obj.setBlockIntervalSrc(reader.readInt());
        obj.setBlockIntervalDst(reader.readInt());
        obj.setMaxAggregation(reader.readInt());
        obj.setDelayLimit(reader.readInt());
        obj.setRelayIdx(reader.readInt());
        obj.setRotateHeight(reader.readLong());
        obj.setRxHeight(reader.readLong());
        obj.setRxHeightSrc(reader.readLong());
        obj.setSackTerm(reader.readInt());
        obj.setSackNext(reader.readLong());
        obj.setSackHeight(reader.readLong());
        obj.setSackSeq(reader.readNullable(BigInteger.class));
        if (reader.beginNullableList()) {
            List<BTPAddress> reachable = new ArrayList<>();
            while(reader.hasNext()) {
                reachable.add(reader.readNullable(BTPAddress.class));
            }
            obj.setReachable(reachable);
            reader.end();
        }
        reader.end();
        return obj;
    }

    public void writeObject(ObjectWriter writer) {
        writer.beginList(16);
        writer.writeNullable(this.getAddr());
        writer.writeNullable(this.getRxSeq());
        writer.writeNullable(this.getTxSeq());
        writer.write(this.getBlockIntervalSrc());
        writer.write(this.getBlockIntervalDst());
        writer.write(this.getMaxAggregation());
        writer.write(this.getDelayLimit());
        writer.write(this.getRelayIdx());
        writer.write(this.getRotateHeight());
        writer.write(this.getRxHeight());
        writer.write(this.getRxHeightSrc());
        writer.write(this.getSackTerm());
        writer.write(this.getSackNext());
        writer.write(this.getSackHeight());
        writer.writeNullable(this.getSackSeq());
        List<BTPAddress> reachable = this.getReachable();
        if (reachable != null) {
            writer.beginNullableList(reachable.size());
            for(BTPAddress v : reachable) {
                writer.write(v);
            }
            writer.end();
        } else {
            writer.writeNull();
        }
        writer.end();
    }

    public static Link fromBytes(byte[] bytes) {
        ObjectReader reader = Context.newByteArrayObjectReader("RLPn", bytes);
        return Link.readObject(reader);
    }

    public byte[] toBytes() {
        ByteArrayObjectWriter writer = Context.newByteArrayObjectWriter("RLPn");
        Link.writeObject(writer, this);
        return writer.toByteArray();
    }
}
